File: src/types/PatronCategory.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "./Locale.js"
import { CultureIdentifier } from "./_Identifier.js"

export const PatronCategory = DB.Entity(import.meta.url, {
  name: "PatronCategory",
  namePlural: "PatronCategories",
  type: () =>
    DB.Object({
      primary_patron_cultures: DB.Required({
        comment:
          "The list of cultures where patrons from this category can be the primary patron of.",
        type: DB.Array(CultureIdentifier(), { minItems: 1, uniqueItems: true }),
      }),
      translations: NestedTranslationMap(
        DB.Required,
        "PatronCategory",
        DB.Object({
          name: DB.Required({
            comment: "The patron category’s name.",
            type: DB.String({ minLength: 1 }),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

File: Sources/OptolithDatabaseSchema/Types/PatronCategory.swift
import FileDB

@Model
public struct PatronCategory {
    /// The list of cultures where patrons from this category can be the primary patron of.
    @MinItems(1)
    @UniqueItems
    @Relationship(Culture.self)
    let primary_patron_cultures: [Culture.ID]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation {  // PatronCategoryTranslation
        /// The patron category’s name.
        @MinLength(1)
        let name: String
    }
}
